python/tank/bootstrap/resolver.py

```diff
@@ -206,172 +282,106 @@ def resolve_shotgun_configuration(
+                        log.debug("Per-user match: %s" % pc)
+
+                else:
+
+                    # alternatively, this is a pipeline configuration record
+                    # which doesn't match directly - typically this is a
+                    # pipeline config record with project id set to None, in this
+                    # case indicating that this configuration can be used for
+                    # *any* project on the site (one config used for all projects).
+                    # this has a lower priority than the exact match above, so if
+                    # a project has specific pipeline configuration specified, this
+                    # always takes precedence.
+                    if pc["code"] == constants.PRIMARY_PIPELINE_CONFIG_NAME:
+                        primary_config_fallback = pc
+                        log.debug("Found primary fallback match: %s" % pc)
+                    else:
+                        user_config_fallback = pc
+                        log.debug("Found per-user fallback match: %s" % pc)
 
             # Now select in order of priority:
-
             if user_config:
                 # A per-user pipeline config for the current project has top priority
                 pipeline_config = user_config
 
+            elif primary_config:
+                # if there is a primary config for our current project, this takes precedence
+                pipeline_config = primary_config
+
             elif user_config_fallback:
                 # if there is a pipeline config for our current user with project field None
                 # that takes precedence
                 pipeline_config = user_config_fallback
 
-            elif primary_config:
-                # if there is a primary config for our current project, this takes precedence
-                pipeline_config = primary_config
-
             elif primary_config_fallback:
                 # Lowest priority - A Primary pipeline configuration with project field None
                 pipeline_config = primary_config_fallback
 
             else:
                 # we may not have any pipeline configuration matches at all:
                 pipeline_config = None
-
-
         else:
-            # there is a fixed pipeline configuration name specified.
-            log.debug("Will use pipeline configuration '%s'" % pipeline_config_name)
+            log.debug("Will use pipeline configuration id '%s'" % pipeline_config_identifier)
 
             log.debug("Requesting pipeline configuration data from Shotgun...")
 
-            pipeline_configs = sg_connection.find(
+            # Fetch the one and only config that matches this id.
+            pipeline_config = sg_connection.find_one(
                 "PipelineConfiguration",
                 [
                     ["project", "is", self._proj_entity_dict],
-                    ["code", "is", pipeline_config_name],
+                    ["id", "is", pipeline_config_identifier],
```

Comment:
I left it in since it should be an error to return the pipeline configuration for the wrong project if the wrong id is passed in. Shall I add a comment to that effect?